File: backend/app/models/scan.py
import enum
import uuid
from datetime import datetime
from typing import TYPE_CHECKING

from sqlalchemy import DateTime, Enum, ForeignKey, func
from sqlalchemy.dialects.postgresql import JSONB
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.db.base_class import Base

if TYPE_CHECKING:
    from app.models.project import Project
    from app.models.target import Target


class ScanStatus(str, enum.Enum):
    CREATED = "created"
    QUEUED = "queued"
    RUNNING = "running"
    DISCOVERY = "discovery"
    FINGERPRINTING = "fingerprinting"
    ANALYZING = "analyzing"
    CORRELATING = "correlating"
    RISK_ANALYSIS = "risk_analysis"
    REPORT_GENERATION = "report_generation"
    COMPLETED = "completed"
    FAILED = "failed"
    CANCELLED = "cancelled"


class Scan(Base):
    __tablename__ = "scans"
    id: Mapped[uuid.UUID] = mapped_column(primary_key=True, default=uuid.uuid4)
    project_id: Mapped[uuid.UUID] = mapped_column(
        ForeignKey("projects.id", ondelete="CASCADE"), nullable=False, index=True
    )
    target_id: Mapped[uuid.UUID] = mapped_column(
        ForeignKey("targets.id", ondelete="CASCADE"), nullable=False, index=True
    )
    status: Mapped[ScanStatus] = mapped_column(
        Enum(
            ScanStatus,
            name="scan_status",
            values_callable=lambda enum_cls: [e.value for e in enum_cls],
        ),
        default=ScanStatus.CREATED,
        nullable=False,
    )
    config: Mapped[dict] = mapped_column(JSONB, default=dict)
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now())
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), onupdate=func.now()
    )
    started_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)
    completed_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)

    project: Mapped["Project"] = relationship(back_populates="scans")
    target: Mapped["Target"] = relationship()

File: frontend/src/lib/types/scan.ts
/**
 * Types mirroring the backend's Scan Pydantic schemas
 * (see backend Task 5 `ScanStatus` enum and Task 6-8 `ScanRead` / `ScanCreate`).
 */

export type ScanStatus =
  | "created"
  | "queued"
  | "running"
  | "discovery"
  | "fingerprinting"
  | "analyzing"
  | "correlating"
  | "risk_analysis"
  | "report_generation"
  | "completed"
  | "failed"
  | "cancelled";

export interface ScanRead {
  id: string;
  project_id: string;
  target_id: string;
  status: ScanStatus;
  config: Record<string, unknown>;
  created_at: string;
  updated_at: string;
  started_at: string | null;
  completed_at: string | null;
  error_message: string | null;
}

export interface ScanCreate {
  project_id: string;
  target_id: string;
  config?: Record<string, unknown>;
}
